pkg/gpu/probe.go

```diff
@@ -0,0 +1,204 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package gpu
+
+import (
+	"fmt"
+	"regexp"
+
+	manager "github.com/DataDog/ebpf-manager"
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/rlimit"
+
+	"github.com/DataDog/datadog-agent/comp/updater/telemetry"
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/gpu/model"
+	ddebpf "github.com/DataDog/datadog-agent/pkg/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/ebpf/bytecode"
+	"github.com/DataDog/datadog-agent/pkg/ebpf/uprobes"
+	"github.com/DataDog/datadog-agent/pkg/util/kernel"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// TODO: Set a minimum kernel version
+var minimumKernelVersion = kernel.VersionCode(5, 0, 0)
+
+const (
+	cudaEventMap      = "cuda_events"
+	cudaAllocCacheMap = "cuda_alloc_cache"
+	gpuAttacherName   = "gpu"
+)
+
+// Probe represents the GPU monitoring probe
+type Probe struct {
+	mgr      *ddebpf.Manager
+	cfg      *Config
+	consumer *CudaEventConsumer
+	attacher *uprobes.UprobeAttacher
+}
+
+// NewProbe starts the GPU monitoring probe
+func NewProbe(cfg *Config, telemetryComponent telemetry.Component) (*Probe, error) {
+	log.Debugf("starting GPU monitoring probe...")
+	kv, err := kernel.HostVersion()
+	if err != nil {
+		return nil, fmt.Errorf("kernel version: %s", err)
+	}
+	if kv < minimumKernelVersion {
+		return nil, fmt.Errorf("minimum kernel version %s not met, read %s", minimumKernelVersion, kv)
+	}
+
+	var probe *Probe
+	filename := "gpu.o"
+	if cfg.BPFDebug {
+		filename = "gpu-debug.o"
+	}
+	err = ddebpf.LoadCOREAsset(filename, func(buf bytecode.AssetReader, opts manager.Options) error {
+		var err error
+		probe, err = startGPUProbe(buf, opts, telemetryComponent, cfg)
+		if err != nil {
+			return fmt.Errorf("cannot start GPU monitoring probe: %s", err)
+		}
+		log.Debugf("started GPU monitoring probe")
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("loading asset: %s", err)
+	}
+
+	return probe, nil
+}
+
+func startGPUProbe(buf bytecode.AssetReader, opts manager.Options, _ telemetry.Component, cfg *Config) (*Probe, error) {
+	mgr := ddebpf.NewManagerWithDefault(&manager.Manager{
+		Maps: []*manager.Map{
+			{Name: cudaAllocCacheMap},
+		}})
+
+	if opts.MapSpecEditors == nil {
+		opts.MapSpecEditors = make(map[string]manager.MapSpecEditor)
+	}
+
+	opts.MapSpecEditors[cudaEventMap] = manager.MapSpecEditor{
+		Type:       ebpf.RingBuf,
+		MaxEntries: 4096,
```

Comment:
I think it should work for now, we shouldn't be seeing a massive amount of activity in this buffer anyways.

Reply:
You should still express it in terms of page size. The kernel requires 1 page to be the minimum buffer size, and the page size is not always `4096`

Reply:
Fixed, with some calculation to ensure it's at least 4096 bytes (I don't think we can get <4K pages, but just in case). Probably in the future we'll want to optimize the size of this buffer once we have more data.